### PR TITLE
RF - Moved tracking.localtrack._local_tracker to DirectionGetter.generate_streamline.

### PR DESCRIPTION
**Pull request by gabknight**
This is the first PR to generalized DIPY's tracking framework.

This PR moved the code from `dipy.tracking.localtrack._local_tracker()` to the `DirectionGetter.generate_streamline()`.  This will allow for `DirectionGetters` to override how streamlines are generated in the local tracking framework. DIPY's `DirectionGetters` select the next tracking based on the current position and previous direction until a stopping criterion is reached. This modification will allow new `DirectionGetters` to use, e.g., the whole sequence of previous directions and positions to select the next tracking direction. 

This doesn't have any effect on the generated streamlines or interfaces. 

The `ParticleFileteringTracking` is not affected by the change. Future PR should modify `PFT` to leave streamline generation to `DirectionGetters` in a similar fashion, such that it can benefit from future overriding `DirectionGetter.generate_streamline`.

### CODE REVIEW

**Comment by gabknight** [comment]
@skoudoro @guaje 

Can you give me feedback on this? I moved the `generate_streamline()` code to the abstract `DirectionGetter` class. It was previously wrongly in `BasePmfDirectionGetter`, it needed to be in the parent class to work `EuDXDirrectionGetter`, which derives directly from `DirectionGetter`. Thanks!

File: dipy/direction/closest_peak_direction_getter.pyx [hunk 1]
```diff
@@ -1,16 +1,24 @@
 import numpy as np
 cimport numpy as cnp
+cimport cython
```
Comment (skoudoro) [inline dipy/direction/closest_peak_direction_getter.pyx:3]: I do not think this import is needed. This is not used

File: dipy/direction/closest_peak_direction_getter.pyx [hunk 2]
```diff
@@ -1,16 +1,24 @@
...
 from dipy.direction.pmf cimport SimplePmfGen, SHCoeffPmfGen
 from dipy.reconst.shm import order_from_ncoef, sph_harm_lookup
 from dipy.tracking.direction_getter cimport DirectionGetter
+from dipy.tracking.stopping_criterion cimport (StreamlineStatus,
+                                               StoppingCriterion,
+                                               TRACKPOINT,
+                                               ENDPOINT,
+                                               OUTSIDEIMAGE,
+                                               INVALIDPOINT,
+                                               PYERROR)
```
Comment (skoudoro) [inline dipy/direction/closest_peak_direction_getter.pyx:17]: I do not think this import is needed. This is not used

File: dipy/direction/closest_peak_direction_getter.pxd [hunk 3]
```diff
@@ -16,10 +18,20 @@ cdef class BasePmfDirectionGetter(DirectionGetter):
...
+    cpdef tuple generate_streamline(self,
+                                    double[::1] seed,
+                                    double[::1] dir,
+                                    double[::1] voxel_size,
+                                    double step_size,
+                                    StoppingCriterion stopping_criterion,
+                                    cnp.float_t[:, :] streamline,
+                                    StreamlineStatus stream_status,
+                                    double use_fixed_step)
+
```
Comment (skoudoro) [inline dipy/direction/closest_peak_direction_getter.pxd:34]: I do not think this is needed. You do not overwrite it and already defined in `DirectionGetter`

File: dipy/direction/probabilistic_direction_getter.pyx [hunk 4]
```diff
@@ -15,6 +15,7 @@ cimport numpy as cnp
 from dipy.direction.closest_peak_direction_getter cimport PmfGenDirectionGetter
 from dipy.direction.peaks import peak_directions, default_sphere
 from dipy.direction.pmf cimport PmfGen, SimplePmfGen, SHCoeffPmfGen
+from dipy.tracking.stopping_criterion cimport StoppingCriterion
```
Comment (skoudoro) [inline dipy/direction/probabilistic_direction_getter.pyx:18]: I do not think this import is needed. This is not used

File: dipy/direction/probabilistic_direction_getter.pyx [hunk 5]
```diff
@@ -32,8 +33,7 @@ cdef class ProbabilisticDirectionGetter(PmfGenDirectionGetter):
         double[:, :] vertices
         dict _adj_matrix
 
-    def __init__(self, pmf_gen, max_angle, sphere=None, pmf_threshold=0.1,
-                 **kwargs):
+    def __init__(self, pmf_gen, max_angle, sphere, pmf_threshold=.1, **kwargs):
```
Comment (skoudoro) [inline dipy/direction/probabilistic_direction_getter.pyx:36]: is the `sphere` mandatory in all `DirectionGetter` ?

File: dipy/direction/probabilistic_direction_getter.pyx [hunk 6]
```diff
@@ -141,8 +141,7 @@ cdef class DeterministicMaximumDirectionGetter(ProbabilisticDirectionGetter):
     function (pmf).
     """
 
-    def __init__(self, pmf_gen, max_angle, sphere=None, pmf_threshold=0.1,
-                 **kwargs):
+    def __init__(self, pmf_gen, max_angle, sphere, pmf_threshold=.1, **kwargs):
```
Comment (skoudoro) [inline dipy/direction/probabilistic_direction_getter.pyx:144]: is the `sphere` mandatory in all `DirectionGetter` ?

File: dipy/tracking/direction_getter.pxd [hunk 7]
```diff
@@ -1,14 +1,28 @@
 
-cimport numpy as np
+from .direction_getter cimport DirectionGetter
```
Comment (skoudoro) [inline dipy/tracking/direction_getter.pxd:2]: I do not think this import is needed. It is strange here

File: dipy/tracking/direction_getter.pxd [hunk 8]
```diff
@@ -1,14 +1,28 @@
...
+from .stopping_criterion cimport (StreamlineStatus, StoppingCriterion)
```
Comment (skoudoro) [inline dipy/tracking/direction_getter.pxd:3]: `from dipy.tracking.stopping_criterion`  instead

File: dipy/tracking/direction_getter.pyx [hunk 9]
```diff
@@ -1,11 +1,141 @@
+cimport cython
+from .direction_getter cimport DirectionGetter
```
Comment (skoudoro) [inline dipy/tracking/direction_getter.pyx:2]: I do not think this import is needed. It is strange here to import itself 😄

File: dipy/tracking/direction_getter.pyx [hunk 10]
```diff
@@ -1,11 +1,141 @@
...
+
     cpdef cnp.ndarray[cnp.float_t, ndim=2] initial_direction(
             self, double[::1] point):
         pass
 
+
+    @cython.boundscheck(False)
+    @cython.wraparound(False)
+    @cython.cdivision(True)
+    cpdef tuple generate_streamline(self,
```
Comment (skoudoro) [inline dipy/tracking/direction_getter.pyx:94]: we should decide on a new name for this function before going further or merging

File: dipy/tracking/localtrack.pyx [hunk 11]
```diff
@@ -86,9 +86,9 @@ cdef void fixed_step(double * point, double * direction, double step_size) nogil
...
         StoppingCriterion sc,
-        cnp.float_t[:] seed_pos,
-        cnp.float_t[:] first_step,
-        cnp.float_t[:] voxel_size,
+        double[::1] seed_pos,
+        double[::1] first_step,
+        double[::1] voxel_size,
         cnp.float_t[:, :] streamline,
         double step_size,
         int fixedstep):
```
Comment (skoudoro) [inline dipy/tracking/localtrack.pyx:94]: `fixedstep` is a bool in the documentation. Can you update the type here or the documentation?

File: dipy/tracking/localtrack.pyx [hunk 12]
```diff
@@ -129,70 +129,23 @@ def local_tracker(
...
-        double point[3]
-        double voxdir[3]
-        void (*step)(double*, double*, double) nogil
-
     if fixedstep:
-        step = fixed_step
+        use_fixed_step = 1
     else:
-        step = step_to_boundary
+        use_fixed_step = 0
```
Comment (skoudoro) [inline dipy/tracking/localtrack.pyx:142]: `use_fixed_step` I do not think this variable is necessary. 

you can use `fixedstep` variable above.

- if `fixedstep` is a bool, you can do `int(fixedstep)` to get your `int`. 
-  if `fixedstep` is a int, you already have your value.

so `use_fixed_step` is a duplicate.

**Review by skoudoro — COMMENTED** [review]
Hi @gabknight,

See below some of my comments.  I agree that `generate_streamlines` should be on `DirectionGetter`. we need to rename `generate_streamlines` from now. it is really confusing with the one in `tracking.py`

Overall, it looks good to me.

File: dipy/direction/closest_peak_direction_getter.pyx (hunk 2 above)
Comment (gabknight) [reply dipy/direction/closest_peak_direction_getter.pyx:17]: yes, thx.

File: dipy/direction/closest_peak_direction_getter.pyx (hunk 1 above)
Comment (gabknight) [reply dipy/direction/closest_peak_direction_getter.pyx:3]: yes

File: dipy/direction/closest_peak_direction_getter.pxd (hunk 3 above)
Comment (gabknight) [reply dipy/direction/closest_peak_direction_getter.pxd:34]: ok, thx.

File: dipy/direction/probabilistic_direction_getter.pyx (hunk 5 above)
Comment (gabknight) [reply dipy/direction/probabilistic_direction_getter.pyx:36]: for all `BasePmfDirectionGetter` yes. There is a default value in the `PmfGenDirectionGetter.from_pmf(.)` and `PmfGenDirectionGetter.from.shcoefs(.)`. If sphere is None, the directionGetters won't work.

File: dipy/direction/probabilistic_direction_getter.pyx (hunk 4 above)
Comment (gabknight) [reply dipy/direction/probabilistic_direction_getter.pyx:18]: yes, removed.

File: dipy/tracking/direction_getter.pxd (hunk 7 above)
Comment (gabknight) [reply dipy/tracking/direction_getter.pxd:2]: yes, it not needed. removed.

File: dipy/tracking/direction_getter.pyx (hunk 9 above)
Comment (gabknight) [reply dipy/tracking/direction_getter.pyx:2]: yes... removed

File: dipy/direction/probabilistic_direction_getter.pyx (hunk 5 above)
Comment (skoudoro) [reply dipy/direction/probabilistic_direction_getter.pyx:36]: make sense, thanks!

File: dipy/tracking/localtrack.pyx (hunk 12 above)
Comment (gabknight) [reply dipy/tracking/localtrack.pyx:142]: yes... forgot to clean this. thx.

File: dipy/tracking/direction_getter.pxd (hunk 8 above)
Comment (gabknight) [reply dipy/tracking/direction_getter.pxd:3]: fixed.

File: dipy/tracking/localtrack.pyx (hunk 11 above)
Comment (gabknight) [reply dipy/tracking/localtrack.pyx:94]: updated. thx.

File: dipy/tracking/direction_getter.pyx (hunk 10 above)
Comment (gabknight) [reply dipy/tracking/direction_getter.pyx:94]: I renamed `dipy.tracking.local_tracking._generate_streamlines(.)` to `dipy.tracking.local_tracking._generate_tractogram(.)`.

Thus, `generate_tractogram` calls its `_tracker(.)`, which internally calls the `DG.generate_streamline()`. 

What do you think?

**Comment by gabknight** [comment]
@skoudoro thx for the feedback. I believe all is fixed. I also suggested changing `_generate_streamlines(.)` to `_generate_tractogram(.)`  in `dipy.tracking.local_tracking.py`.  

I don't need to add deprecated warnings for internal function changes, do I?

File: dipy/tracking/direction_getter.pyx (hunk 10 above)
Comment (skoudoro) [reply dipy/tracking/direction_getter.pyx:94]: I personally love the idea. make more sense to me

File: dipy/direction/closest_peak_direction_getter.pyx [hunk 13]
```diff
@@ -58,6 +57,73 @@ cdef int closest_peak(np.ndarray[np.float_t, ndim=2] peak_dirs,
...
+        The direction along which the step should be taken.
+    point : c-pointer to double[3]
+        The tracking point which will be updated by this function.
+    step_size : double
+        The size of step in units of direction.
+
+    """
+    for i in range(3):
+        point[i] += direction[i] * step_size
+
```
Comment (skoudoro) [inline dipy/direction/closest_peak_direction_getter.pyx:126]: I do not understand why you need this block in `close_peak_direction_getter.pyx`. it is not used.

**Review by skoudoro — COMMENTED** [review]
Hi @gabknight,

Thank you for the update. 

I like the strategy to rename `_generate_streamlines(.)` to `_generate_tractogram(.)` in `dipy.tracking.local_tracking.py`. it makes more sense to me.

Apart from that, there are some functions in `closest_peaks` not used so not necessary?

> I don't need to add deprecated warnings for internal function changes, do I?

No you don't, no needs

File: dipy/direction/closest_peak_direction_getter.pyx (hunk 13 above)
Comment (gabknight) [reply dipy/direction/closest_peak_direction_getter.pyx]: we don't need it. This should have been removed when we moved `generate_streamline(.)` to the parent class `DirectionGetter`. thx.

Comment (skoudoro) [reply dipy/direction/closest_peak_direction_getter.pyx]: Ok, Great!  This PR is ready for me. 

There are 2-3 things that can be improved (unnecessary copy) but we can do that in a new PR.

**Comment by skoudoro** [comment]
I think this PR is ready to go!  I will wait until tomorrow evening for additional comments. Feel free to review this PR!

**Comment by skoudoro** [comment]
Thank you @gabknight!

**Pull request merged**